quickwit/quickwit-indexing/failpoints/mod.rs

```diff
@@ -307,20 +307,27 @@ async fn test_merge_executor_controlled_directory_kill_switch() -> anyhow::Resul
     // aborted not by the actor framework, before the message is being processed.
     //
     // To do so, we
-    // - pause the actor right before the merge operation
-    // - send the message
-    // - wait 500ms to make sure the test has reached the "pause" point
+    // - set two barrier so the actor pauses right upon entering the process_merge function
+    // - send the merge message
+    // - wait on the first barrier to ensure that the actor has reached the process_merge function
     // - kill the universe
-    // - unpause
+    // - wait and release the second barrier so the actor can continue processing the merge message
     //
     // Before the controlled directory, the merge operation would have continued until it
     // finished, taking hundreds of millisecs to terminate.
-    fail::cfg("before-merge-split", "pause").unwrap();
+    let before_universe_kill = Arc::new(Barrier::new(2));
+    let after_universe_kill = Arc::new(Barrier::new(2));
+    let before_universe_kill_clone = before_universe_kill.clone();
+    let after_universe_kill_clone = after_universe_kill.clone();
+    fail::cfg_callback("before-merge-split", move || {
+        before_universe_kill_clone.wait();
+        after_universe_kill_clone.wait();
+    })
+    .unwrap();
     merge_executor_mailbox.send_message(merge_scratch).await?;
-
-    tokio::time::sleep(Duration::from_millis(10)).await;
```

Comment:
Alternatively you can put an arbitrary amount of time here. We work with simulated time.

Reply:
I couldn't quite figure out how to make this sleep simulated. Even if I replace `tokio::time::sleep()` with `universe.sleep()` this sleep is still real. The envelope of the merge_scratch message created on the line above is not getting dropped until `merge_executor_handle.join`, which happens after sleep and while this envelope is alive `no_advance_time_guard` in it prevents the simulated sleep from happening. So, while putting 500ms here does indeed help these are real 500ms.

Reply:
ah right because we pause within a handler. that makes sense